railgun/internal/manager/controllerdef.go

```diff
@@ -199,49 +199,60 @@ func setupControllers(logger logr.Logger, mgr manager.Manager, proxy proxy.Proxy
+		kongClusterPluginGVR := schema.GroupVersionResource{
+			Group:    konghqcomv1.SchemeGroupVersion.Group,
+			Version:  konghqcomv1.SchemeGroupVersion.Version,
+			Resource: "kongclusterplugins",
+		}
+
+		if ctrlutils.CRDExists(mgr.GetClient(), kongClusterPluginGVR) {
+			logger.Info("kongclusterplugins.configuration.konghq.com v1beta1 CRD available on cluster.")
+			controller := ControllerDef{
+				IsEnabled: &c.KongClusterPluginEnabled,
+				Controller: &configuration.KongV1KongClusterPluginReconciler{
+					Client:           mgr.GetClient(),
+					Log:              ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
+					Scheme:           mgr.GetScheme(),
+					Proxy:            proxy,
+					IngressClassName: c.IngressClassName,
+				},
+			}
+			controllers = append(controllers, controller)
+		} else {
+			logger.Info(`kongclusterplugins.configuration.konghq.com v1beta1 CRD not available on cluster.
+		Disabling KongClusterPlugin controller`)
 		}
-		controllers = append(controllers, controller)
 	} else {
-		logger.Info(`kongclusterplugins.configuration.konghq.com v1beta1 CRD not available on cluster.
+		logger.Info(`kong cluster plugin is disabled.
 		Disabling KongClusterPlugin controller`)
 	}
+	if c.KnativeIngressEnabled == util.EnablementStatusEnabled {
+		knativeGVR := schema.GroupVersionResource{
+			Group:    knativev1alpha1.SchemeGroupVersion.Group,
+			Version:  knativev1alpha1.SchemeGroupVersion.Version,
+			Resource: "ingresses",
+		}
 
-	knativeGVR := schema.GroupVersionResource{
-		Group:    knativev1alpha1.SchemeGroupVersion.Group,
-		Version:  knativev1alpha1.SchemeGroupVersion.Version,
-		Resource: "ingresses",
-	}
-	if ctrlutils.CRDExists(mgr.GetClient(), knativeGVR) {
-		logger.Info("ingresses.networking.internal.knative.dev v1alpha1 CRD available on cluster.")
-		controller := ControllerDef{
-			IsEnabled: &c.KnativeIngressEnabled,
-			Controller: &configuration.Knativev1alpha1IngressReconciler{
-				Client:           mgr.GetClient(),
-				Log:              ctrl.Log.WithName("controllers").WithName("Ingress").WithName("KnativeV1Alpha1"),
-				Scheme:           mgr.GetScheme(),
-				Proxy:            proxy,
-				IngressClassName: c.IngressClassName,
-			},
+		if ctrlutils.CRDExists(mgr.GetClient(), knativeGVR) {
+			logger.Info("ingresses.networking.internal.knative.dev v1alpha1 CRD available on cluster.")
+			controller := ControllerDef{
+				IsEnabled: &c.KnativeIngressEnabled,
+				Controller: &configuration.Knativev1alpha1IngressReconciler{
+					Client:           mgr.GetClient(),
+					Log:              ctrl.Log.WithName("controllers").WithName("Ingress").WithName("KnativeV1Alpha1"),
+					Scheme:           mgr.GetScheme(),
+					Proxy:            proxy,
+					IngressClassName: c.IngressClassName,
+				},
+			}
+			controllers = append(controllers, controller)
+		} else {
+			logger.Info(`ingresses.networking.internal.knative.dev v1alpha1 CRD not available on cluster.
+		Disabling Knative controller`)
```

Comment:
Similar to above, I'm again wondering if we should fail if the caller specifically enabled the controller but the CRD is not enabled. Fail fast seems to be the preferable mechanism here, rather than expecting the caller to launch the controller and immediate read the logs to see if any controllers failed.

```suggestion
			message := fmt.Sprintf("%s CRD not available on the cluster", knativeGVR.String())
			if c.KnativeIngressEnabled == util.EnablementStatusEnabled {
				return nil, fmt.Errorf(message)
			}
			logger.Infof("%s: disabling controller", message)
```

Reply:
Agree if controller is enabled but KIC could not support. Should  fails to start.